### PR TITLE
Reduce Location instance allocations.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Addresses about .5% of all allocs in a simply typing/lightbulb scenario:

image

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@333fred @RikkiGibson ptal.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceFieldSymbol.cs [hunk 1]
```diff
@@ -211,33 +211,19 @@ public sealed override string Name
         }
 
         internal override LexicalSortKey GetLexicalSortKey()
-        {
-            return new LexicalSortKey(_location, this.DeclaringCompilation);
-        }
+            => new LexicalSortKey(_syntaxReference, this.DeclaringCompilation);
+
+        public override Location TryGetFirstLocation()
+            => _syntaxReference.SyntaxTree.GetLocation(_locationSpan);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceFieldSymbol.cs:217]: Is there any risk that certain access patterns here could actually make us allocate more of these?

**Review by RikkiGibson — COMMENTED** [review]
Done review pass. Just want to make sure I understand the impact+motivation of the change.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceFieldSymbol.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceFieldSymbol.cs:217]: I thought the location is actually used for local variables in order to report "use before declaration" errors. Maybe it is already working in terms of spans.

It seems very plausible that almost all of the time there is no error reported on a local, and if there is, it is usually 1 error. Thanks!

**Review by RikkiGibson — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
@jaredpar ptal

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceFieldSymbol.cs [hunk 2]
```diff
@@ -211,33 +211,19 @@ public sealed override string Name
         }
 
         internal override LexicalSortKey GetLexicalSortKey()
-        {
-            return new LexicalSortKey(_location, this.DeclaringCompilation);
-        }
+            => new LexicalSortKey(_syntaxReference, this.DeclaringCompilation);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceFieldSymbol.cs:214]: I wonder if we should think about deleting the `Location, Compilation` constructor, and force instanciations to use SyntaxRefs directly. Maybe a followup?

**Review by 333fred — APPROVED** [review]

**Pull request merged**